### DISCUSSION TITLE
Running optuna across different machines

### DISCUSSION THREAD

**Discussion by dln22** [optuna/optuna#4114] [Q&A] [answered]
Hello,

I have a small question about my optuna hyper parameter optimization setup. I am running optuna on the grid, where the jobs are launched at the same time on different machines. The main communication point for all the jobs is the study.db file. I found that if I do not change the seed on the TPE sampler on each machine then there are trials with the same parameters being tried. So my current setup can be schematically defined as follows:
1. Initialise an empty optuna study db file before launching any jobs on the grid. This is to avoid a race condition for creating the database file.
2. Launch optuna hpo on the grid, where we load the study and change the study sampler seed, each job on the grid will have a different seed. So I am doing something like this:
```python
    # this would load the study from a previously created file
    study = optuna.create_study(
        direction=direction, storage=storage, study_name=study_name, load_if_exists=load_if_exists
    )
    # change the TPE sampler seed
    study.sampler = optuna.samplers.TPESampler(consider_prior=True, n_startup_trials=10, seed=opts.seed)
    study.optimize(...)
```
3. Run the finalize step, where I save all the trials info to a csv file

My question is, is it ok to have parallel runs, each on different machine, which share the same study, but have different sampler or rather different sampler seeds ? Or is there a better way to do it? I found that without changing the seed, the first trial on each machine is the same. 

My goal is to optimize a very expensive model, thus I might go for a setup, where there would be 1-5 trials per machine, and where I would have 30-100 machines in parallel. I only wonder if my setup is correct for the job.
Many thanks!

**Comment by not522** [comment, marked as answer]
In distributed optimization, `seed=None` is preferred. It means randomly initializing the random seed.

**Reply by dln22** [reply]
Thank you very much for the suggestion. I added it to my code. I also found that if the study already exists and we want to simply add more trials to it, then re-creating the sampler and attaching it to the study is incorrect, as it leads to the same trials being sampled, e.g.:

```python
# MY OLD INCORRECT CODE
 study = optuna.create_study(
      direction=direction, storage=storage, study_name=study_name, load_if_exists=load_if_exists
  )
  # if study exists already, this would cause the same trials being sampled
  study.sampler = optuna.samplers.TPESampler(consider_prior=True, n_startup_trials=10, seed=opts.seed)
```

```python
# CORRECTED CODE
 study = optuna.create_study(
      direction=direction, storage=storage, study_name=study_name, load_if_exists=load_if_exists
  )

# add a function that checks the study existance
if not study_exists(<args>):
  study.sampler = optuna.samplers.TPESampler(consider_prior=True, n_startup_trials=10, seed=opts.seed)
```

Given the tests I have done so far, it seems my setup is doing ok. Thanks!

**Reply by not522** [reply]
Your 2nd implementation is incorrect because the default sampler (i.e., `TPESampler` with the default arguments) is used. If you want to customize the sampler, you need to specify it in all studies. (Note: the sampler is not saved in the storage.)
